clarity-ext-scripts/clarity_ext_scripts/covid/parse_pcr_execution.py

```diff
@@ -2,33 +2,75 @@
 
 import pandas as pd
 from datetime import datetime
+from clarity_ext.domain.validation import UsageError
 
-CT_HEADER = u"Cт"
+CT_HEADER_OLD = u"Cт"
 
 
 class ParsePcrExecution(object):
     def __init__(self, context):
         self.context = context
 
     def execute(self):
-        file_name = "Result file"
-        f = self.context.local_shared_file(file_name, mode="rb")
-        data = pd.read_excel(f, 'Results', index_col=0, skiprows=7, encoding='utf-8')
-        rt_pcr_analyze_service = RTPCRAnalyseService(self.context)
+        if not self._has_instrument_udf():
+            raise UsageError("The udf 'Instrument Used' must be filled in before running this script")
+        file_handle = "Result file"
+        parser = self._instantiate_parser()
+        parser.parse(file_handle)
+
+    def _instantiate_parser(self):
+        if self.instrument == 'qPCR ABI 7500':
+            raise UsageError('Parsing qPCR ABI 7500 is not implemented')
+        elif self.instrument == 'Quant Studio 7':
+            parser = Quant7Parser(self.context)
+        else:
+            raise UsageError("The instrument in 'Instrument Used' is not recognized: {}"
+                             .format(self.instrument))
+        return parser
+
+    @property
+    def instrument(self):
+        return self.context.current_step.udf_instrument_used
+
+    def _has_instrument_udf(self):
+        try:
+            _ = self.instrument
+        except AttributeError:
+            return False
+
+        return True
 
-        for _, output in self.context.all_analytes:
-            entry = data.loc[output.well.alpha_num_key]
-            target_name = entry["Target Name"]
 
+class Quant7Parser(object):
+    def __init__(self, context):
+        self.context = context
+
+    def _determine_start_row(self, file_handle):
+        file_stream = self.context.local_shared_file(file_handle, mode="rb")
+        data = pd.read_excel(file_stream, 'Results', index_col=None, encoding='utf-8')
+        header_row_entry = data[(data[0] == 'Well') & (data[1] == 'Well Position')].index
+        header_row_index = header_row_entry.values[0]
+        return header_row_index
+
+    def parse(self, file_handle):
+        header_row_index = self._determine_start_row(file_handle)
+        file_stream = self.context.local_shared_file(file_handle, mode="rb")
+        data = pd.read_excel(file_stream, 'Results', index_col=None, skiprows=header_row_index, encoding='utf-8')
+        for _, output in self.context.all_analytes:
+            fam_row = data.loc[(data['Sample Name'] == 'sample1') & (data['Reporter'] == 'FAM')]
```

Comment:
Won't this only pick up rows where the `Sample Name` matches `sample1`? I guess we will need other samples as well?

Reply:
quack, I'm a bit stressed right now.

Reply:
Np. We will make sure everything comes out alright together.